lib/scholar/metrics/distance.ex

```diff
@@ -240,24 +349,62 @@ defmodule Scholar.Metrics.Distance do
       iex> y = Nx.tensor([1, 2, 3])
       iex> Scholar.Metrics.Distance.cosine(x, y)
       ** (ArgumentError) expected input shapes to be equal, got {2} != {3}
+
+      iex> x = Nx.tensor([[1, 2, 3], [0, 0, 0], [5, 2, 4]])
+      iex> y = Nx.tensor([[1, 5, 2], [2, 4, 1], [0, 0, 0]])
+      iex> Scholar.Metrics.Distance.cosine(x, y, axes: [1])
+      #Nx.Tensor<
+        f32[3]
+        [0.1704850196838379, 1.0, 1.0]
+      >
   """
-  @spec cosine(Nx.t(), Nx.t()) :: Nx.t()
-  defn cosine(x, y) do
+  @spec cosine(Nx.t(), Nx.t(), keyword()) :: Nx.t()
+  defn cosine(x, y, opts \\ []) do
+    # Detect very small values that could lead to surprising
+    # results and numerical stability issues. Every value smaller
+    # than `cutoff` is considered small
+    cutoff = 10 * 2.220446049250313e-16
     assert_same_shape!(x, y)
-    norm_x = Nx.LinAlg.norm(x)
-    norm_y = Nx.LinAlg.norm(y)
-
-    cond do
-      norm_x == 0.0 and norm_y == 0.0 ->
-        0.0
 
-      norm_x == 0.0 or norm_y == 0.0 ->
-        1.0
+    opts = keyword!(opts, [:axes])
 
-      true ->
-        numerator = Nx.dot(x, y)
-        denominator = norm_x * norm_y
-        1.0 - numerator / denominator
+    if Nx.all(x == y) do
+      0.0
```

Comment:
Why do we need this special case? In particular note that this currently ignores `:axes`.